src/sig_stfl/xmss/LICENSE

```diff
@@ -0,0 +1,12 @@
+## License
+
+This XMSS reference implementation is Copyright (c) 2024 SandboxAQ and licensed under both the [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0.txt) and [MIT License](LICENSE-MIT).
+
+Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.
+
+This XMSS reference implementation is based on the [XMSS reference implementation written by Andreas Hülsing and Joost Rijneveld](https://github.com/XMSS/xmss-reference#license) provided under the CC0 1.0 Universal Public Domain Dedication.
+
+
+## Disclaimer
+
+The software and documentation are provided "as is" and SandboxAQ hereby disclaims all warranties, whether express, implied, statutory, or otherwise. SandboxAQ specifically disclaims, without limitation, all implied warranties of merchantability, fitness for a particular purpose, title, and non-infringement, and all warranties arising from course of dealing, usage, or trade practice. SandboxAQ makes no warranty of any kind that the software and documentation, or any products or results of the use thereof, will meet any person's requirements, operate without interruption, achieve any intended result, be compatible or work with any software, system or other services, or be secure, accurate, complete, free of harmful code, or error-free.
```

Comment:
This disclaimer alone now makes me reject all of this XMSS code for inclusion in `liboqs`.

Rationale:
- No other code contains company-specific responsibility disclaimers
- It has been added after the notion of adding XMSS has been challenged on safety grounds (https://github.com/open-quantum-safe/liboqs/pull/1650#issuecomment-1893957110), so Sandbox AQ seems to agree with the risks stated; so why should other OQS contributors (let alone users, not reading all single LICENSE files) take on this risk?

My proposal to move forward with this is to relegate this code to a specific "absolutely-no-responsibility-here" sub project (that also "absolves" _all_ OQS contributors of any consequences of copying, using, deploying, etc. this code). This would allow `liboqs` to continue evolution to "product utility" and this code to be experimented with.

Reply:
The MIT License already contains a disclaimer: https://en.wikipedia.org/wiki/MIT_License

Reply:
Sigh -- I have to apologize again: I have been misled then by [the stated goal of the PR "to maximize the adoption"](https://github.com/open-quantum-safe/liboqs/pull/1662#issuecomment-1887861704). But you are right, @dstebila, by understanding that this code adheres to _both_ MIT AND Apache2, the disclaimer of MIT already "protects" everyone (from the consequences of  code labelled ["borderline irresponsible"](https://github.com/open-quantum-safe/liboqs/pull/1650#issuecomment-1893957110)). Makes me wonder why the Disclaimer has been added in the first place -- unless we have a different understanding of "AND" and "OR", so let me ask explicitly:

Do we all agree that the "AND" clause means that this code will be _excluded_ from the library for anyone building `liboqs` only under Apache2 (as soon as #1514 is implemented -- which _I_ consider a feature maximizing adoption of `liboqs` as the "algorithm amalgamation library" that it is)? Particularly asking @ducnguyen-sb as this would arguably _not_ maximize adoption of this code (any entity not accepting MIT then will not be able to use this code).

Lastly, why not move all these combined license terms into one LICENSE file? Still "feels" not quite right... The question of why the CCO license of the original code is no longer valid also remains open.

Reply:
I think this would be where we could get some advice from LF folks on how to sort through this. I'll see if Hart can point someone towards this.

Reply:
I've asked the folks at the Linux Foundation who are experts on open source licensing to address this.  I'll get back in touch with everyone once we hear back.  Thanks!

Reply:
Hi @baentsch , sorry for the late reply. 
I speak for myself here. 
Regards to "AND" or "OR" license, I can change it according to what `liboqs` wants. 
Regards to the question why not continue CC0 license, I found the issue in liboqs: https://github.com/open-quantum-safe/liboqs/issues/1437, some may find this license is an obstacle. 

Thus, I search around, read wikipedia regards to CC0: https://www.wikiwand.com/en/Creative_Commons_license#Zero_/_public_domain , of course I am not a lawyer, among those words, I think it can be re-licensed. 
My poor attempt at searching if CC0 can be relicensed: https://opensource.stackexchange.com/questions/1386/may-someone-else-apply-a-license-to-cc0-material, and I find the answer is "yes". 
Of course, the Internet can be wrong and many on the Internet are not lawyers.

Okay, let's talk about the Disclaimer in the license file: 

1. MIT: feel free to use it, don't sue us (first reminder)
2. Apache 2.0: feel free to use it, don't sue us (second reminder)
3. A Disclaimer: feel free to use it, please don't sue us (TRIPLE REMINDER)

I don't think that the `Disclaimer` is a bad practice, see [`aws-lc`](https://github.com/aws/aws-lc) project licensed under Apache 2.0 OR ISC, you will see at the end their License file, it has Disclaimer in all UPPERCASE. 

Again, I think any company want to protect themselves from being sue. So having a redundant is an affirmative way to do so. 
If you think this Disclaimer is redundant, I can remove it. 

Or I can revert the license to CC0 (fine for us), and perhaps hope that it won't be an obstacle someday?

Reply:
Just to make it super clear that other than Apache 2.0 AND MIT, no addition license is attached.

Reply:
> Or I can revert the license to CC0 (fine for us)

Putting everything under CC0 to me sounds like the most "correct" solution: If this is the original authors' license, how can we simply change it? 

> and perhaps hope that it won't be an obstacle someday?

This may well be: AFAIK, LinuxFoundation is not supportive of open source in all meanings of the term, so I will not comment on this license issue any more, but leave this to the entities interested in protecting their commercial interests as per [the above](https://github.com/open-quantum-safe/liboqs/pull/1650#discussion_r1462260229).

Reply:
I'm not sure either whether we can re-license CC0 to MIT unilaterally. I'll email the LF lawyers to ask them. Otherwise we would want to go back to the authors of the original implementation to ask if they are willing to relicense under MIT.  The desire to not include CC0 code here comes from past issues that have been raised (such as #1437) that CC) is problematic for adopters, such as Fedora.

Reply:
> I'm not sure either whether we can re-license CC0 to MIT unilaterally. I'll email the LF lawyers to ask them. 

After checking the original code, I'd consider that unnecessary: Even if the LF lawyers answer it'd be OK (for them), I'd consider this deeply unethical from a FOSS perspective and would thus strongly object such move: I just checked a single [sample file](https://github.com/XMSS/xmss-reference/blob/master/xmss.c) and it's 99% unchanged included in this PR (though amended).

> Otherwise we would want to go back to the authors of the original implementation to ask if they are willing to relicense under MIT. The desire to not include CC0 code here comes from past issues that have been raised (such as https://github.com/open-quantum-safe/liboqs/issues/1437) that CC) is problematic for adopters, such as Fedora.

This is also something I would find unnecessary for several reasons:
1) @ducnguyen-sb already stated "I can revert the license to CC0 (fine for us)" -- so I'd be _very_ glad about that.
2) The reference code is very old and I'm not certain it'd be possible to get hold of all authors to agree to such change of license.
3) The entity that complained about CC0 is IBM/Redhat (Fedora) -- the very same company labeling (some features of) this code [borderline irresponsible](https://github.com/open-quantum-safe/liboqs/pull/1650#issuecomment-1893957110) ... which I take to mean they'd not over-eagerly integrate this :) But let's check: @simo5: Can you comment on this question: Would you want to distribute this code (without setting OQS_ALLOW_XMSS_KEY_AND_SIG_GEN of course)? Would this decision change if it were CC0 or MIT licensed?

Reply:
Just a comment here, while IBM wholly owns Red Hat, we are different (and largely independent) entities, please never assume that we have unity of representation or even intent.

That said we (Red Hat) will not compile the siggen parts for sure, but CC0 has issues in general even as source code distribution, and it is not a problem just for us, it's just that we tend to be the ones that performs license due diligence early compared to other players.

In general CC0 is problematic *for software* and you should reconsider using it at all.

Reply:
Thanks for the feedback.

> please never assume that we have unity of representation or even intent.

ACK. No insult intended (though not sure to whom :).

> In general CC0 is problematic for software and you should reconsider using it at all.

OK -- can you point to some treatise explaining why this is so?

Reply:
I think, but do not quote me on this because I have not asked directly, that CC0 by explicitly disclaiming any implied patent grant is a risk of submarine patent attack.

I honestly do not want to go into a debate on whether this is a problem or not, just stating what the perceived risk is which makes CC0 undesirable for code that should be free to use by all, including entities with commercial interests.

Reply:
> CC0 by explicitly disclaiming any implied patent grant is a risk of submarine patent attack.

Thanks. I now (finally, I'm slow...) get it. A more verbose statement of about the same risk:

> There is a clause in the text of the license CC0 which explicitly states that the license does not affect any patent or trademark rights that may be used in the application. The possibility of influence through patents is considered a potential threat, so licenses that do not explicitly allow the use of patents or do not waive patents are considered non-FOSS

(from https://blog.desdelinux.net/en/in-fedora-they-plan-to-stop-offering-various-spins-and-software-under-the-cc0-license/)

So I now also don't like CC0 any more. Doesn't make our problem easier given that the code in this PR (IMO) _is_ CC0.

So final question to you, @simo5 : Would you package it (without sig/gen) if we'd somehow get it (changed) to be MIT&Apache2?

Reply:
With my Red Hat fedora on: It being packaged in RHEL largely depends on industry needs. If they become widely used then we will have no choice but to ship support for XMSS and LMS.

Reply:
I can't give you a final answer at this stage because, as Hubert notes, license is just one of the factors.

We have disabled/removed unwanted (weak, broken, patent encumbered, etc...) algorithms in cryptography libraries in the past, so we'll make a determination based on multiple factors, but certainly license is one of them.